distributed/worker.py

```diff
@@ -647,30 +648,31 @@ def _register_with_scheduler(self):
 
     @gen.coroutine
     def heartbeat(self):
-        if not self.heartbeat_active:
-            self.heartbeat_active = True
-            logger.debug("Heartbeat: %s" % self.address)
-            try:
-                start = time()
-                response = yield self.scheduler.heartbeat_worker(
-                    address=self.contact_address,
-                    now=time(),
-                    metrics=self.get_metrics()
-                )
-                end = time()
-                middle = (start + end) / 2
-
-                if response['status'] == 'missing':
-                    yield self._register_with_scheduler()
-                    return
-                self.scheduler_delay = response['time'] - middle
-                self.periodic_callbacks['heartbeat'].callback_time = response['heartbeat-interval'] * 1000
-            except CommClosedError:
-                logger.warning("Heartbeat to scheduler failed")
-            finally:
-                self.heartbeat_active = False
-        else:
-            logger.debug("Heartbeat skipped: channel busy")
+        with (yield self.heartbeat_lock.acquire()):
+            if not self.heartbeat_active:
+                self.heartbeat_active = True
+                logger.debug("Heartbeat: %s" % self.address)
+                try:
+                    start = time()
+                    response = yield self.scheduler.heartbeat_worker(
+                        address=self.contact_address,
+                        now=time(),
+                        metrics=self.get_metrics()
+                    )
+                    end = time()
+                    middle = (start + end) / 2
+
+                    if response['status'] == 'missing':
+                        yield self._register_with_scheduler()
+                        return
+                    self.scheduler_delay = response['time'] - middle
+                    self.periodic_callbacks['heartbeat'].callback_time = response['heartbeat-interval'] * 1000
+                except CommClosedError:
+                    logger.warning("Heartbeat to scheduler failed")
+                finally:
+                    self.heartbeat_active = False
+            else:
+                logger.debug("Heartbeat skipped: channel busy")
```

Comment:
I was finding that while this method has a variable that it checks to see if it another thread is currently performing a heartbeat check, and thus issues the "channel busy" message and skipping the heartbeat check. 

There is a race condition, i.e. two threads may be have have checked the if branch, and found that a heart beat check is necessary. This was found  as there was multiple concurrent heart-beat messages on the scheduler from the same worker. 

Ideally the register with scheduler should be forced to be single threaded resource with appropriate tests to prevent concurrent operation as well as tests to verify that the registration is still necessary. 

I found that not only the periodic heart beat would could call register with scheduler in multiple threads, and if the network connection is broken, that handler would call register with scheduler, thus you could end up with multiple threads concurrently trying to register with the scheduler.

Then to top it off, if one thread successfully registered with scheduler, then the other thread would fail and throw an exception, which would disable the re-registration of the heart beat scheduler, as at the start of the register with scheduler, would disable the period heart beat, and the thrown exception would prevent re-registration.